src/main/java/edu/hm/hafner/coverage/parser/CloverParser.java

```diff
@@ -0,0 +1,170 @@
+                    }
+                }
+            }
+            handleEmptyResults(fileName, log);
+            return new ModuleNode("empty");
+        } catch (XMLStreamException exception) {
+            throw new SecureXmlParserFactory.ParsingException(exception);
+        }
+    }
+
+    @CanIgnoreReturnValue
+    private ModuleNode readCoverage(final XMLEventReader reader, final ModuleNode root, String fileName) throws XMLStreamException {
+        while (reader.hasNext()) {
+            XMLEvent event = reader.nextEvent();
+
+            if (event.isStartElement()) {
+                var startElement = event.asStartElement();
+                if (PACKAGE.equals(startElement.getName())) {
+                    readPackage(reader, root, startElement, fileName);
+                }
+            } else if (event.isEndElement()) {
+                var endElement = event.asEndElement();
+                if (COVERAGE.equals((endElement.getName()))) {
+                    return root;
+                }
+            }
+        }
+        throw createEofException(fileName);
+    }
+
+    @CanIgnoreReturnValue
+    private PackageNode readPackage(final XMLEventReader reader, final ModuleNode root, final StartElement packageElement, String fileName)
+            throws XMLStreamException {
+        var packageName = getValueOf(packageElement, NAME);
+        var packageNode = root.findOrCreatePackageNode(packageName);
+
+        while (reader.hasNext()) {
+            XMLEvent event = reader.nextEvent();
+
+            if (event.isStartElement()) {
+                var startElement = event.asStartElement();
+                if (FILE.equals(startElement.getName())) {
+                    readFile(reader, packageNode, startElement);
+                }
+            } else if (event.isEndElement()) {
+                var endElement = event.asEndElement();
+                if (PACKAGE.equals((endElement.getName()))) {
+                    return packageNode;
+                }
+            }
+        }
+        throw createEofException(fileName);
+    }
+
+    @CanIgnoreReturnValue
+    private FileNode readFile(final XMLEventReader reader, final PackageNode packageNode, final StartElement fileElement) throws XMLStreamException {
+        String fileName = getValueOf(fileElement, NAME);
+        String className = fileName.substring(0, fileName.lastIndexOf("."));
+        String filePath = getValueOf(fileElement, PATH);
+        // TODO check if filePath is relative or absolute path (latter needs to remove everything except current workspace)
+        var fileNode = packageNode.findOrCreateFileNode(fileName, TreeString.valueOf(filePath));
+        var classNode = packageNode.findOrCreateClassNode(className);
+        while (reader.hasNext()) {
+            XMLEvent event = reader.nextEvent();
+            if (event.isStartElement()) {
+                var e = event.asStartElement();
+                if (METRICS.equals(e.getName())) {
```

Comment:
It looks like you are ignoring the metrics of the top level elements, is this intended? In JaCoCo the sum of the children might be different than the value in the parent because some counters cannot be mapped to classes or methods. So it might make sense to map these values as well (and assert them in the test).

Reply:
Yeah I wasn't sure how to handle the case where the top level elements' metrics didn't match up with the metrics by consolidating the children's metrics, so I was ignoring them and using the children's metrics as the source of truth. But I can add the top level metrics to their respective nodes.

Reply:
I started with your existing approach in my other parsers as well, but then some users complained that the overall coverage numbers (that are based on the sums in the leaves) do not match with the aggregated numbers in the report. So I changed the behavior and now use the values in the report and only use the calculated aggregation if no aggregation is available. (See https://github.com/jenkinsci/coverage-model/blob/bdad7860e846fedd8a4e32313e2bdb18891f2f0c/src/main/java/edu/hm/hafner/coverage/Metric.java#L246-L253)